From 3c46c8a6dd54f215f335cb6dcef25cb8e936eed9 Mon Sep 17 00:00:00 2001
From: Chen Minqiang <ptpt52@gmail.com>
Date: Wed, 5 Aug 2026 02:21:17 +0800
Subject: [PATCH] mtd: spi-nor: allow force unlocking via DT property

Some SPI NOR flash chips (such as generic or unlisted chips used in vendor
devices like Tenda AX12L Pro) have Block Protection (BP) bits set in the
Status Register by bootloaders or factory settings, locking flash blocks.

Because vendors frequently switch between various generic SPI NOR flash
chips ("Flash Lottery"), it is impractical to upstream explicit chip ID
flags (SNOR_F_HAS_LOCK) for every possible generic chip variant.

This patch introduces support for the "linux,force-sr-unlock" Device Tree
property:
1. In spi_nor_init_fixup_flags(), set the SNOR_F_HAS_LOCK flag on the flash
   instance if "linux,force-sr-unlock" is present in the flash DT node. This
   allows spi_nor_late_init_params() to automatically populate default
   locking_ops without altering swp.c.
2. In spi_nor_init(), trigger spi_nor_try_unlock_all() if "linux,force-sr-unlock"
   is set, invoking Linux kernel's native spi_nor_unlock() mechanism.

Signed-off-by: Chen Minqiang <ptpt52@gmail.com>
---
 drivers/mtd/spi-nor/core.c | 5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -2924,6 +2924,9 @@ static void spi_nor_init_fixup_flags(str
 {
 	const u8 fixup_flags = nor->info->fixup_flags;
 
+	if (of_property_read_bool(spi_nor_get_flash_node(nor), "linux,force-sr-unlock"))
+		nor->flags |= SNOR_F_HAS_LOCK;
+
 	if (fixup_flags & SPI_NOR_4B_OPCODES)
 		nor->flags |= SNOR_F_4B_OPCODES;
 
@@ -3261,7 +3264,8 @@ static int spi_nor_init(struct spi_nor *
 	 */
 	if (IS_ENABLED(CONFIG_MTD_SPI_NOR_SWP_DISABLE) ||
 	    (IS_ENABLED(CONFIG_MTD_SPI_NOR_SWP_DISABLE_ON_VOLATILE) &&
-	     nor->flags & SNOR_F_SWP_IS_VOLATILE))
+	     nor->flags & SNOR_F_SWP_IS_VOLATILE) ||
+	    of_property_read_bool(spi_nor_get_flash_node(nor), "linux,force-sr-unlock"))
 		spi_nor_try_unlock_all(nor);
 
 	if (nor->addr_nbytes == 4 &&
